### PR TITLE
Inline Enum.into with bitstring implementation

### PR DESCRIPTION
**Pull request by tsubery**
Replaces recursive io list implementation with a flat list
implementation.

According to my benchmarks it improves performance by 6x-35x according
to collection length.
See the following test script and results:

```Elixir
def compare do
 thousand = Enum.map(1..1000, &to_string/1)
 Benchee.run(%{
   "before 1k" => fn ->
     Enum.into(thousand, "")
   end,
   "after 1k" => fn ->
     Enum.into(thousand, [])
     |> IO.iodata_to_binary
   end
 },time: 2)

 million = Enum.map(1..1000_000, &to_string/1)
 Benchee.run(%{
   "before 1m" => fn ->
     Enum.into(million, "")
   end,
   "after 1m" => fn ->
     Enum.into(million, [])
     |> IO.iodata_to_binary
   end
 },time: 10)

 million_x_100 = Enum.map(1..100_000_000, &to_string/1)
 Benchee.run(%{
   "before 100m" => fn ->
     Enum.into(million_x_100, "")
   end,
   "after 100m" => fn ->
     Enum.into(million_x_100, [])
     |> IO.iodata_to_binary
   end
 }, time: 30)
end
```

Results:

```Shell
Name                ips        average  deviation         median
after 1k       160.45 K        6.23 μs    ±70.31%        6.00 μs
before 1k       25.69 K       38.93 μs    ±18.08%       37.00 μs

Comparison:
after 1k       160.45 K
before 1k       25.69 K - 6.25x slower

Name                ips        average  deviation         median
after 1m         124.98        8.00 ms     ±4.01%        7.97 ms
before 1m         16.17       61.83 ms    ±28.17%       51.69 ms

Comparison:
after 1m         124.98
before 1m         16.17 - 7.73x slower

Name                  ips        average  deviation         median
after 100m           0.83         1.21 s   ±140.25%         0.85 s
before 100m        0.0241        41.55 s     ±0.00%        41.55 s

Comparison:
after 100m           0.83
before 100m        0.0241 - 34.47x slower
```

### CODE REVIEW

**Comment by ericmj** [comment]
@tsubery Can you include benchmarks of the new implementation since it differs from your reference in the comparisons? I ran a few basic tests myself and could not get a difference significant enough to convince me the new implementation is more performant.

**Comment by tsubery** [comment]
Your are correct, the benchmark did not reflect the code performance. This is because ```into: []``` is actually short circuited and bypasses Collectable protocol. I added another fix and a more comprehensive benchmark to the PR.
The ~20% improvement of the Collectable implementation is arguably insignificant. Seems harmless improvement since it is doesn't add complexity.

**Comment by josevalim** [comment]
@tsubery can we keep only the second commit please? I still don't see any noticiable difference for the collectable one (not even 20%).

**Comment by tsubery** [comment]
These are the results I get with [this benchmark](https://gist.github.com/tsubery/303dbdbcd65f8de49b197029e3265978)
Seems like 50% slower on the bigger lists. Maybe it has to do with GC pressure.
I'm curious if it is something with my machine or the bench setup.

##### With input 100m #####
Name                       ips        average  deviation         median
skip collectable          0.97         1.03 s    ±23.54%         0.98 s
before                  0.0831        12.04 s     ±0.00%        12.04 s
flat_list               0.0551        18.16 s     ±0.00%        18.16 s

Comparison:
skip collectable          0.97
before                  0.0831 - 11.72x slower
flat_list               0.0551 - 17.68x slower

##### With input 1k #####
Name                       ips        average  deviation         median
skip collectable      108.83 K        9.19 μs   ±281.46%        9.00 μs
before                 20.33 K       49.20 μs   ±102.15%       40.00 μs
flat_list              17.23 K       58.03 μs  ±2336.83%       56.00 μs

Comparison:
skip collectable      108.83 K
before                 20.33 K - 5.35x slower
flat_list              17.23 K - 6.32x slower

##### With input 1m #####
Name                       ips        average  deviation         median
skip collectable        113.70        8.79 ms     ±3.59%        8.74 ms
flat_list                15.97       62.62 ms    ±21.45%       53.85 ms
before                   14.84       67.40 ms    ±31.98%       55.43 ms

Comparison:
skip collectable        113.70
flat_list                15.97 - 7.12x slower
before                   14.84 - 7.66x slower

**Comment by josevalim** [comment]
@tsubery are the benchmarks warming up the code before running it?

**Comment by josevalim** [comment]
Nevermind, I could see that it is. :)

**Comment by josevalim** [comment]
@tsubery I need to investigate because I can't see those changes locally. But I think the PR is currently good as is, so let's at least merge the current inlining!

Thank you!

**Comment by tsubery** [comment]
Created a [repo](https://github.com/tsubery/misc_bench/) with the whole project to lock down benchee version if that helps investigation

File: lib/elixir/lib/enum.ex [hunk 1]
```diff
@@ -1099,6 +1099,14 @@ defmodule Enum do
     collectable ++ to_list(enumerable)
   end
 
+  def into(enumerable, collectable) when is_binary(collectable) do
+    postfix =
+      enumerable
+      |> to_list
+      |> IO.iodata_to_binary
+    collectable <> postfix
```
Comment (ericmj) [inline lib/elixir/lib/enum.ex]: The iodata to binary conversion can be moved down here to possibly avoid one extra allocation `IO.iodata_to_binary([collectable | postfix])`.

**Comment by michalmuskala** [comment]
An important thing to notice is that this changes the behaviour of `Enum.into/2` with a binary.

Before `Enum.into([1], "")` would fail, now it will produce `<<1>>`

**Comment by tsubery** [comment]
@michalmuskala That's a good point. Both behaviours seem reasonable to me as long as it's consistent between Enum.into and Collectable implementations. Having said that, this commit might introduce some unexpected behavior.

**Comment by josevalim** [comment]
@tsubery we need to stick with the behaviour of `for ..., into: ""` which is optimized by the VM and only allow binaries, so we need to fix the behaviour @michalmuskala commented.

**Comment by tsubery** [comment]
Having another pass over the enumerable to verify all elements are binaries takes away most of the performance advantage introduced by this PR.

**Pull request closed without merging**